cosim/emulator/src/vbridge_impl.cc

```diff
@@ -0,0 +1,465 @@
+}
+
+
+void VBridgeImpl::reset() {
+  top.clock = 0;
+  top.reset = 1;
+  top.resetVector = 0x1000;
+  top.eval();
+  tfp.dump(0);
+
+  // posedge
+  top.clock = 1;
+  top.eval();
+  tfp.dump(1);
+
+
+  // negedge
+  top.reset = 0;
+  top.clock = 0;
+  top.eval();
+  tfp.dump(2);
+  // posedge
+  LOG(INFO) << fmt::format("Reset compeleted, now we start");
+  top.reset = 0;
+  top.clock = 1;
+  top.eval();
+  tfp.dump(3);
+  ctx.time(2);
+}
+
+VBridgeImpl::~VBridgeImpl() {
+  terminate_simulator();
+}
+
+void VBridgeImpl::configure_simulator(int argc, char **argv) {
+  ctx.commandArgs(argc, argv);
+}
+
+void VBridgeImpl::init_spike() {
+  auto state = proc.get_state();
+  LOG(INFO) << fmt::format("Spike reset misa={:08X}", state->misa->read());
+  LOG(INFO) << fmt::format("Spike reset mstatus={:08X}", state->mstatus->read());
+  // load binary to reset_vector
+  sim.load(bin, ebin, reset_vector);
+}
+
+SpikeEvent *find_se_to_issue();
+void VBridgeImpl::init_simulator() {
+  Verilated::traceEverOn(true);
+  top.trace(&tfp, 99);
+  tfp.open(wave.c_str());
+  _cycles = timeout;
+}
+
+void VBridgeImpl::terminate_simulator() {
+  if (tfp.isOpen()) {
+    tfp.close();
+  }
+  top.final();
+}
+
+uint64_t VBridgeImpl::get_t() {
+  return ctx.time();
+}
+
+uint8_t VBridgeImpl::load(uint64_t address) {
+  return *sim.addr_to_mem(address);
+}
+
+/* For TL Acquire
+ * Record mem info in se.block when init spike event;
+ * receive tl Acquire, find corresponding SE in queue. store se.block info in AcquireBanks;set AcquireBanks.remaining as true;
+ * return: drive D channel with AcquireBanks
+ * */
+void VBridgeImpl::run() {
+
+  init_spike();
+  // sim.load(bin, reset_vector);
+  init_simulator();
+  reset();
+
+  // start loop
+  // queue size must > for trap insn to pop correct
+  while (true) {
+    loop_until_se_queue_full();
+    while (to_rtl_queue.size() > 1) {
+      return_fetch_response();
+
+      top.clock = 1;
+      top.eval();
+
+      top.memory_0_a_ready = 1;
+      top.memory_0_c_ready = 1;
+      top.memory_0_e_ready = 1;
+      // negedge
+
+      receive_tl_req();
+      top.clock = 0;
+      top.eval();
+      tfp.dump(2 * ctx.time());
+      ctx.timeInc(1);
+
+
+      // posedge, update registers
+      top.clock = 1;
+      top.eval();
+      tfp.dump(2 * ctx.time() - 1);
+
+
+      // when RTL write back
+      if (top.rootp->DUT__DOT__ldut__DOT__tile__DOT__core__DOT__wb_valid) {
+        uint64_t pc = top.rootp->DUT__DOT__ldut__DOT__tile__DOT__core__DOT__wb_reg_pc;
+        LOG(INFO) << fmt::format("RTL write back insn {:08X} ", pc);
+        // Check rf write
+        // todo: use rf_valid
+        if (top.rootp->DUT__DOT__ldut__DOT__tile__DOT__core__DOT____Vcellinp__rf_ext__W0_en) {
+          record_rf_access();
+        }
+        // commit spike event
+        bool commit = false;
+
+        for (auto se_iter = to_rtl_queue.rbegin(); se_iter != to_rtl_queue.rend(); se_iter++) {
+          if (se_iter->pc == pc) {
```

Comment:
Is this compare-by-pc strategy sound enough, since a loop may issue multiple instruction of the same pc?

Reply:
If there are multiple event with the same pc in the queue, we will find the one which entered earlier.  Considering Rocket is an in-order core, it will work.